c2rust-analyze/src/rewrite/expr/distribute.rs

```diff
@@ -0,0 +1,126 @@
+/// MIR rewrites.
+///
+/// Using the example from `unlower`:
+///
+/// ```text
+/// bb0[5]: Terminator { source_info: ..., kind: _4 = f(move _5) -> [return: bb1, unwind: bb2] }
+///   []: StoreIntoLocal, `f(y)`
+///   [Rvalue]: Expr, `f(y)`
+///   [Rvalue, CallArg(0)]: Expr, `y`
+/// ```
+///
+/// A MIR rewrite on `bb0[5]` `[Rvalue, CallArg(0)]` would be attached to the MIR
+/// `Expr` `y`, and a rewrite on `bb0[5]` `[Rvalue]` would be attached to `f(y)`.
+/// A MIR rewrite on `bb0[5]` `[]` (i.e. on the call terminator itself) would
+/// result in an error, since there is no good place in the HIR to attach such a
+/// rewrite.
+pub fn distribute<'tcx>(
+    tcx: TyCtxt<'tcx>,
+    unlower_map: BTreeMap<(Location, Vec<SubLoc>), MirOrigin>,
+    mir_rewrites: HashMap<Location, Vec<MirRewrite>>,
+) -> HashMap<HirId, Vec<mir_op::RewriteKind>> {
+    let mut info_map = HashMap::<HirId, Vec<RewriteInfo>>::new();
+
+    for (loc, mir_rws) in mir_rewrites {
+        for mir_rw in mir_rws {
+            let mut key = (loc, mir_rw.sub_loc);
+
+            let mut origin = unlower_map.get(&key);
+            if origin.is_none() && matches!(key.1.last(), Some(&SubLoc::RvalueOperand(0))) {
+                // Hack: try without the `RvalueOperand(0)` at the end.
+                // TODO: remove SubLoc::RvalueOperand from mir_op
+                key.1.pop();
+                origin = unlower_map.get(&key);
+                if origin.is_none() {
+                    key.1.push(SubLoc::RvalueOperand(0));
+                }
+            }
+            let mut origin = match origin {
+                Some(x) => x,
+                None => {
+                    error!("unlower_map has no origin for {:?}", key);
+                    continue;
+                }
+            };
+
+            if origin.desc == MirOriginDesc::StoreIntoLocal && key.1.len() == 0 {
+                // Hack: try with an extra `Rvalue` in the key.
+                // TODO: add SubLoc::Rvalue in mir_op ptr::offset handling
+                key.1.push(SubLoc::Rvalue);
+                match unlower_map.get(&key) {
+                    Some(o) if o.desc == MirOriginDesc::Expr => {
+                        origin = o;
+                    }
+                    _ => {
+                        key.1.pop();
+                    }
+                }
+            }
+
+            if origin.desc != MirOriginDesc::Expr {
+                error!(
+                    "can't distribute rewrites onto {:?} origin {:?}\n\
+                        key = {:?}\n\
+                        mir_rw = {:?}",
+                    origin.desc, origin, key, mir_rw.kind
+                );
+            }
+
+            info_map
+                .entry(origin.hir_id)
+                .or_default()
+                .push(RewriteInfo {
+                    rw: mir_rw.kind,
+                    loc,
+                    sub_loc: key.1,
+                    desc: origin.desc,
+                });
```

Comment:
I think this might be clearer if written as grouping (i.e. `.into_group_map()` from `itertools`).

Reply:
It looks like this would require restructuring the whole loop into a big iterator chain, which IMO is generally less readable than a standard for loop.